File: 78.py
class Solution:
    def __init__(self):
        self.ans = []

    def gen_sets(self, curr, idx, nums):
        if idx == len(nums):
            self.ans.append(curr)
            return
        # either add the elt to curr and call again or don't add the elt to curr and call again
        self.gen_sets([nums[idx]] + curr, idx + 1, nums)
        self.gen_sets(curr, idx + 1, nums)

    def subsets(self, nums: list[int]) -> list[list[int]]:
        self.gen_sets([nums[0]], 1, nums)
        self.gen_sets([], 1, nums)
        return self.ans


print(Solution().subsets([1, 2, 3]))

File: 78.cpp
#include <algorithm>
#include <deque>
#include <iostream>
#include <map>
#include <set>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;
class Solution {
  vector<vector<int>> ans;

public:
  void gen_subsets(vector<int> &nums, vector<int> &curr, int idx) {
    if (idx == nums.size()) {
      ans.push_back(curr);
      return;
    }
    curr.push_back(nums[idx]);
    gen_subsets(nums, curr, idx + 1);
    curr.pop_back();
    gen_subsets(nums, curr, idx + 1);
  }
  vector<vector<int>> subsets(vector<int> &nums) {
    this->ans = {};
    vector<int> curr = {nums[0]};
    gen_subsets(nums, curr, 1);
    curr.pop_back();
    gen_subsets(nums, curr, 1);
    return ans;
  }
};
int main() {}